### PR TITLE
Add NealsFunnel E2E test case and repro f42811c regression

### LINKED ISSUE #32: E2E test measuring K-S against non-trivial target log density (GMM or Neals funnel)

### PR DESCRIPTION
**Pull request by feynmanliang**
This test fails after f42811c  but passes on its parent 4156c07bd8dea79cae06ef58f4737d2f04b2ebdc

Related to #30 

Closes #32

This change is [Reviewable](https://reviewable.io/reviews/stefanwebb/flowtorch/31)

### CODE REVIEW

File: tests/test_bijector.py [hunk 1]
```diff
@@ -1,17 +1,95 @@
...
+    )
+    opt = torch.optim.Adam(params.parameters(), lr=1e-2)
+    num_elbo_mc_samples = 100
+    for _ in range(300):
+        z0 = tdist.base_dist.rsample(sample_shape=(num_elbo_mc_samples,))
+        zk = flow._forward(z0, params)
+        ldj = flow._log_abs_det_jacobian(z0, zk, params)
+
+        neg_elbo = -nf.log_prob(zk).sum()
+        neg_elbo += tdist.base_dist.log_prob(z0).sum() - ldj.sum()
```
Comment (feynmanliang) [inline tests/test_bijector.py:78]: When I used `tdist.log_prob` I ran into errors `backward`sing the same graph twice, and even after `retain_graph=True` I was getting errors trying to `backward`s through a mutated graph.
I'll create a MVP to repro

**Comment by stefanwebb** [comment]
Looks like we're getting an import error: `ModuleNotFoundError: No module named 'scipy'`

I'm looking into this separately by plotting the output of samples from the distribution/training loop you provided and will report back!

I think this would make a good example for: https://flowtorch.ai/user/start :)

**Comment by feynmanliang** [comment]
Yes we probably need to add it to TEST_REQUIRES; will do tomorrow.

**Comment by stefanwebb** [comment]
Okay, I've discovered a few things:

* It is surprisingly hard to learn a 2D Neal's Funnel, especially with a single affine bijection
* I think the test `test_neals_funnel_vi` isn't that meaningful since it uses so few MC samples to learn the distribution, `100`, and then only `20` samples for the Kolmogorov-Smirnov statistic.
* There was a bug in the `._log_abs_det_jacobian` of  `AffineAutoregressive`
* Learning with the new initialization doesn't seem to work unless I decrease the scale of the columns in the weight matrices of `DenseAutoregressive` by an order of magnitude. Then it works better, I think, than the standard Xavier initialization!

Making these changes, this is about the best solution I could learn:

trying to learn neal's funnel

**Comment by stefanwebb** [comment]
Also, it seems like a terrible idea empirically to rescale the output of `DenseAutoregressive` (the mean and log_scale) in `AffineAutoregressive` so I've removed this...

**Comment by stefanwebb** [comment]
I've got this test to pass! I think we should discuss more whether it's a good test

**Pull request merged**